src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -160,58 +160,168 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         };
 
         let suggest_restriction =
-            |generics: &hir::Generics<'_>, msg, err: &mut DiagnosticBuilder<'_>| {
+            |generics: &hir::Generics<'_>,
+             msg,
+             err: &mut DiagnosticBuilder<'_>,
+             fn_sig: Option<&hir::FnSig<'_>>| {
+                // Type parameter needs more bounds. The trivial case is `T` `where T: Bound`, but
+                // it can also be an `impl Trait` param that needs to be decomposed to a type
+                // param for cleaner code.
                 let span = generics.where_clause.span_for_predicates_or_empty_place();
                 if !span.from_expansion() && span.desugaring_kind().is_none() {
-                    err.span_suggestion(
-                        generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi(),
-                        &format!("consider further restricting {}", msg),
-                        format!(
-                            "{} {} ",
-                            if !generics.where_clause.predicates.is_empty() {
-                                ","
-                            } else {
-                                " where"
+                    // Given `fn foo(t: impl Trait)` where `Trait` requires assoc type `A`...
+                    if let Some((name, fn_sig)) = fn_sig.and_then(|sig| {
+                        projection.and_then(|p| {
+                            // Shenanigans to get the `Trait` from the `impl Trait`.
+                            match p.self_ty().kind {
+                                ty::Param(param) if param.name.as_str().starts_with("impl ") => {
+                                    let n = param.name.as_str();
```

Comment:
```suggestion
                                    let trait_str = param.name.as_str().strip_prefix("impl").trim_start();
```

Reply:
Also, there has to be a cleaner, less stringy way to extract this info... :( cc @eddyb

Reply:
Would it be possible to use the `Span` of the parameter? At the very least that would replace the `param.name.as_str()`. But you might need to get it from some contextual information, since `ty::Param` itself doesn't have it.

Or maybe even better, search in `self.param_env` for bounds on `p.self_ty()`, and find the min/max point along their spans... oh, but `ParamEnv` doesn't have `Span`s, ugh.

Reply:
Applied the suggest. Agree that there should be a better way to do this, but I couldn't find one. It seemed to me that we need to turn `impl Trait` into a more first-class member of `rustc` in order to avoid these kind of things.

Reply:
> It seemed to me that we need to turn `impl Trait` into a more first-class member of `rustc` in order to avoid these kind of things.

That's not exactly true, the problem here is that parameters don't know their *identity*, only an index, and the Chalk integration seems to strongly want to move even further towards indexes only. So what you're missing is contextual information to let you dig deeper into what *defined* this `ty::Param`.